Import libraries

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('train.csv')

In [3]:
df.head()

channelGrouping      date  \
0  Organic Search  20160902   
1  Organic Search  20160902   
2      Affiliates  20160902   
3  Organic Search  20160902   
4  Organic Search  20160902   

                                              device        fullVisitorId  \
0  {"browser": "Chrome", "browserVersion": "not a...  9674781571160116268   
1  {"browser": "Chrome", "browserVersion": "not a...  8590648239310839049   
2  {"browser": "Chrome", "browserVersion": "not a...  9772828344252850532   
3  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
4  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   

                                          geoNetwork  \
0  {"continent": "Asia", "subContinent": "Southea...   
1  {"continent": "Europe", "subContinent": "Easte...   
2  {"continent": "Americas", "subContinent": "Sou...   
3  {"continent": "Americas", "subContinent": "Nor...   
4  {"continent": "Americas", "subContinent": "Nor...   

                        sessionId  socialEngagementType  \
0  9674781571160116268_1472804607  Not Socially Engaged   
1  8590648239310839049_1472835928  Not Socially Engaged   
2  9772828344252850532_1472856802  Not Socially Engaged   
3  1350700416054916432_1472879649  Not Socially Engaged   
4  1350700416054916432_1472829671  Not Socially Engaged   

                                              totals  \
0  {"visits": "1", "hits": "1", "pageviews": "1",...   
1  {"visits": "1", "hits": "1", "pageviews": "1",...   
2  {"visits": "1", "hits": "1", "pageviews": "1",...   
3  {"visits": "1", "hits": "1", "pageviews": "1",...   
4  {"visits": "1", "hits": "5", "pageviews": "4",...   

                                       trafficSource     visitId  visitNumber  \
0  {"campaign": "(not set)", "source": "google", ...  1472804607            1   
1  {"campaign": "(not set)", "source": "google", ...  1472835928            1   
2  {"campaign": "Data Share Promo", "source": "Pa...  1472856802            1   
3  {"campaign": "(not set)", "source": "google", ...  1472879649            2   
4  {"campaign": "(not set)", "source": "google", ...  1472829671            1   

   visitStartTime  
0      1472804607  
1      1472835928  
2      1472856802  
3      1472879649  
4      1472829671

Tamanho do dataset

In [4]:
df.shape

(12283, 12)

verificao da quantidade de sessoes unicas, um usuario pode ter mais de uma visita no site 

In [5]:
len(df['fullVisitorId'].unique())

9996

Verificacao do tipo das colunas, object = string, int64 = inteiro, uint64 = inteiro sem sinal

In [6]:
df.dtypes

channelGrouping         object
date                     int64
device                  object
fullVisitorId           uint64
geoNetwork              object
sessionId               object
socialEngagementType    object
totals                  object
trafficSource           object
visitId                  int64
visitNumber              int64
visitStartTime           int64
dtype: object

Recarregando o dataframe com conversoes das colunas date,fullvisitorId,VisitId para nao perder os zeros a esquerda

In [7]:
df = pd.read_csv('train.csv',dtype={'date':object,'fullVisitorId':object,'VisitId':object})

In [8]:
df.dtypes

channelGrouping         object
date                    object
device                  object
fullVisitorId           object
geoNetwork              object
sessionId               object
socialEngagementType    object
totals                  object
trafficSource           object
visitId                  int64
visitNumber              int64
visitStartTime           int64
dtype: object

Trabalhando com dicionarios = json chave valor

In [9]:
df['device'].iloc[0]

'{"browser": "Chrome", "browserVersion": "not available in demo dataset", "browserSize": "not available in demo dataset", "operatingSystem": "Windows", "operatingSystemVersion": "not available in demo dataset", "isMobile": false, "mobileDeviceBranding": "not available in demo dataset", "mobileDeviceModel": "not available in demo dataset", "mobileInputSelector": "not available in demo dataset", "mobileDeviceInfo": "not available in demo dataset", "mobileDeviceMarketingName": "not available in demo dataset", "flashVersion": "not available in demo dataset", "language": "not available in demo dataset", "screenColors": "not available in demo dataset", "screenResolution": "not available in demo dataset", "deviceCategory": "desktop"}'

Verificamos a cima a quantidade de metadados nen sempre nosso algoritmo de ml vai conseguir absorver todas as informacoes, entao e nescessario, separar em colunas

In [10]:
type(df['device'].iloc[0])

str

In [11]:
import json

In [12]:
type(json.loads(df['device'].iloc[0]))

dict

In [13]:
json.loads(df['device'].iloc[0])

{'browser': 'Chrome',
 'browserVersion': 'not available in demo dataset',
 'browserSize': 'not available in demo dataset',
 'operatingSystem': 'Windows',
 'operatingSystemVersion': 'not available in demo dataset',
 'isMobile': False,
 'mobileDeviceBranding': 'not available in demo dataset',
 'mobileDeviceModel': 'not available in demo dataset',
 'mobileInputSelector': 'not available in demo dataset',
 'mobileDeviceInfo': 'not available in demo dataset',
 'mobileDeviceMarketingName': 'not available in demo dataset',
 'flashVersion': 'not available in demo dataset',
 'language': 'not available in demo dataset',
 'screenColors': 'not available in demo dataset',
 'screenResolution': 'not available in demo dataset',
 'deviceCategory': 'desktop'}

In [14]:
pd.DataFrame([json.loads(linha) for linha in df.device])

browser                 browserVersion  \
0               Chrome  not available in demo dataset   
1               Chrome  not available in demo dataset   
2               Chrome  not available in demo dataset   
3               Safari  not available in demo dataset   
4               Safari  not available in demo dataset   
...                ...                            ...   
12278           Chrome  not available in demo dataset   
12279           Chrome  not available in demo dataset   
12280           Chrome  not available in demo dataset   
12281           Chrome  not available in demo dataset   
12282  Android Webview  not available in demo dataset   

                         browserSize operatingSystem  \
0      not available in demo dataset         Windows   
1      not available in demo dataset       Macintosh   
2      not available in demo dataset       Macintosh   
3      not available in demo dataset             iOS   
4      not available in demo dataset             iOS   
...                              ...             ...   
12278  not available in demo dataset         Windows   
12279  not available in demo dataset         Android   
12280  not available in demo dataset           Linux   
12281  not available in demo dataset         Android   
12282  not available in demo dataset         Android   

              operatingSystemVersion  isMobile           mobileDeviceBranding  \
0      not available in demo dataset     False  not available in demo dataset   
1      not available in demo dataset     False  not available in demo dataset   
2      not available in demo dataset     False  not available in demo dataset   
3      not available in demo dataset      True  not available in demo dataset   
4      not available in demo dataset      True  not available in demo dataset   
...                              ...       ...                            ...   
12278  not available in demo dataset     False  not available in demo dataset   
12279  not available in demo dataset      True  not available in demo dataset   
12280  not available in demo dataset     False  not available in demo dataset   
12281  not available in demo dataset      True  not available in demo dataset   
12282  not available in demo dataset      True  not available in demo dataset   

                   mobileDeviceModel            mobileInputSelector  \
0      not available in demo dataset  not available in demo dataset   
1      not available in demo dataset  not available in demo dataset   
2      not available in demo dataset  not available in demo dataset   
3      not available in demo dataset  not available in demo dataset   
4      not available in demo dataset  not available in demo dataset   
...                              ...                            ...   
12278  not available in demo dataset  not available in demo dataset   
12279  not available in demo dataset  not available in demo dataset   
12280  not available in demo dataset  not available in demo dataset   
12281  not available in demo dataset  not available in demo dataset   
12282  not available in demo dataset  not available in demo dataset   

                    mobileDeviceInfo      mobileDeviceMarketingName  \
0      not available in demo dataset  not available in demo dataset   
1      not available in demo dataset  not available in demo dataset   
2      not available in demo dataset  not available in demo dataset   
3      not available in demo dataset  not available in demo dataset   
4      not available in demo dataset  not available in demo dataset   
...                              ...                            ...   
12278  not available in demo dataset  not available in demo dataset   
12279  not available in demo dataset  not available in demo dataset   
12280  not available in demo dataset  not available in demo dataset   
12281  not available in demo dataset  not available in demo dataset   
12282  not available in demo dataset  not availabl

In [15]:
dicionarios = ['device','geoNetwork','trafficSource','totals']

In [16]:
for coluna in dicionarios:
    df = df.join(pd.DataFrame([json.loads(linha) for linha in df[coluna]]))

In [17]:
df.head()

channelGrouping      date  \
0  Organic Search  20160902   
1  Organic Search  20160902   
2      Affiliates  20160902   
3  Organic Search  20160902   
4  Organic Search  20160902   

                                              device        fullVisitorId  \
0  {"browser": "Chrome", "browserVersion": "not a...  9674781571160116268   
1  {"browser": "Chrome", "browserVersion": "not a...  8590648239310839049   
2  {"browser": "Chrome", "browserVersion": "not a...  9772828344252850532   
3  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
4  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   

                                          geoNetwork  \
0  {"continent": "Asia", "subContinent": "Southea...   
1  {"continent": "Europe", "subContinent": "Easte...   
2  {"continent": "Americas", "subContinent": "Sou...   
3  {"continent": "Americas", "subContinent": "Nor...   
4  {"continent": "Americas", "subContinent": "Nor...   

                        sessionId  socialEngagementType  \
0  9674781571160116268_1472804607  Not Socially Engaged   
1  8590648239310839049_1472835928  Not Socially Engaged   
2  9772828344252850532_1472856802  Not Socially Engaged   
3  1350700416054916432_1472879649  Not Socially Engaged   
4  1350700416054916432_1472829671  Not Socially Engaged   

                                              totals  \
0  {"visits": "1", "hits": "1", "pageviews": "1",...   
1  {"visits": "1", "hits": "1", "pageviews": "1",...   
2  {"visits": "1", "hits": "1", "pageviews": "1",...   
3  {"visits": "1", "hits": "1", "pageviews": "1",...   
4  {"visits": "1", "hits": "5", "pageviews": "4",...   

                                       trafficSource     visitId  ...  \
0  {"campaign": "(not set)", "source": "google", ...  1472804607  ...   
1  {"campaign": "(not set)", "source": "google", ...  1472835928  ...   
2  {"campaign": "Data Share Promo", "source": "Pa...  1472856802  ...   
3  {"campaign": "(not set)", "source": "google", ...  1472879649  ...   
4  {"campaign": "(not set)", "source": "google", ...  1472829671  ...   

   isTrueDirect  referralPath adContent campaignCode visits hits pageviews  \
0           NaN           NaN       NaN          NaN      1    1         1   
1           NaN           NaN       NaN          NaN      1    1         1   
2           NaN           NaN       NaN          NaN      1    1         1   
3          True           NaN       NaN          NaN      1    1         1   
4           NaN           NaN       NaN          NaN      1    5         4   

   bounces newVisits transactionRevenue  
0        1         1                NaN  
1        1         1                NaN  
2        1         1                NaN  
3        1       NaN                NaN  
4      NaN         1                NaN  

[5 rows x 54 columns]

In [18]:
df.shape

(12283, 54)

Depois de tirar as informacoes agora podemos excluir as colunas 

In [21]:
df.drop(dicionarios, axis=1, inplace=True)

In [22]:
df.head()

channelGrouping      date        fullVisitorId  \
0  Organic Search  20160902  9674781571160116268   
1  Organic Search  20160902  8590648239310839049   
2      Affiliates  20160902  9772828344252850532   
3  Organic Search  20160902  1350700416054916432   
4  Organic Search  20160902  1350700416054916432   

                        sessionId  socialEngagementType     visitId  \
0  9674781571160116268_1472804607  Not Socially Engaged  1472804607   
1  8590648239310839049_1472835928  Not Socially Engaged  1472835928   
2  9772828344252850532_1472856802  Not Socially Engaged  1472856802   
3  1350700416054916432_1472879649  Not Socially Engaged  1472879649   
4  1350700416054916432_1472829671  Not Socially Engaged  1472829671   

   visitNumber  visitStartTime browser                 browserVersion  ...  \
0            1      1472804607  Chrome  not available in demo dataset  ...   
1            1      1472835928  Chrome  not available in demo dataset  ...   
2            1      1472856802  Chrome  not available in demo dataset  ...   
3            2      1472879649  Safari  not available in demo dataset  ...   
4            1      1472829671  Safari  not available in demo dataset  ...   

  isTrueDirect referralPath adContent  campaignCode visits hits pageviews  \
0          NaN          NaN       NaN           NaN      1    1         1   
1          NaN          NaN       NaN           NaN      1    1         1   
2          NaN          NaN       NaN           NaN      1    1         1   
3         True          NaN       NaN           NaN      1    1         1   
4          NaN          NaN       NaN           NaN      1    5         4   

  bounces newVisits transactionRevenue  
0       1         1                NaN  
1       1         1                NaN  
2       1         1                NaN  
3       1       NaN                NaN  
4     NaN         1                NaN  

[5 rows x 50 columns]

Limpando a base de dados

In [23]:
df.columns

Index(['channelGrouping', 'date', 'fullVisitorId', 'sessionId',
       'socialEngagementType', 'visitId', 'visitNumber', 'visitStartTime',
       'browser', 'browserVersion', 'browserSize', 'operatingSystem',
       'operatingSystemVersion', 'isMobile', 'mobileDeviceBranding',
       'mobileDeviceModel', 'mobileInputSelector', 'mobileDeviceInfo',
       'mobileDeviceMarketingName', 'flashVersion', 'language', 'screenColors',
       'screenResolution', 'deviceCategory', 'continent', 'subContinent',
       'country', 'region', 'metro', 'city', 'cityId', 'networkDomain',
       'latitude', 'longitude', 'networkLocation', 'campaign', 'source',
       'medium', 'keyword', 'adwordsClickInfo', 'isTrueDirect', 'referralPath',
       'adContent', 'campaignCode', 'visits', 'hits', 'pageviews', 'bounces',
       'newVisits', 'transactionRevenue'],
      dtype='object')

In [24]:
len(df['channelGrouping'].unique())

8

In [29]:
df.drop('adwordsClickInfo', axis=1, inplace=True)

In [30]:
for coluna in df.columns:
    print(coluna+ ':' + str(len(df[coluna].unique())))

channelGrouping:8
date:366
fullVisitorId:9997
sessionId:12276
socialEngagementType:1
visitId:12272
visitNumber:50
visitStartTime:12279
browser:22
browserVersion:1
browserSize:1
operatingSystem:12
operatingSystemVersion:1
isMobile:2
mobileDeviceBranding:1
mobileDeviceModel:1
mobileInputSelector:1
mobileDeviceInfo:1
mobileDeviceMarketingName:1
flashVersion:1
language:1
screenColors:1
screenResolution:1
deviceCategory:3
continent:6
subContinent:22
country:145
region:212
metro:53
city:296
cityId:1
networkDomain:1747
latitude:1
longitude:1
networkLocation:1
campaign:7
source:76
medium:7
keyword:129
isTrueDirect:2
referralPath:198
adContent:16
campaignCode:2
visits:1
hits:96
pageviews:75
bounces:2
newVisits:2
transactionRevenue:147


Avaliando a quantida de valores unicos para cada coluna, um coluna que tem um valor unico para todas as linhas nao traz nada de imortante para o nosso modelo

In [31]:
coluna_na = []

In [32]:
for coluna in df.columns:
    print(coluna+ ':' + str(len(df[coluna].unique())))
    if len(df[coluna].unique()) ==1:
        coluna_na.append(coluna)

channelGrouping:8
date:366
fullVisitorId:9997
sessionId:12276
socialEngagementType:1
visitId:12272
visitNumber:50
visitStartTime:12279
browser:22
browserVersion:1
browserSize:1
operatingSystem:12
operatingSystemVersion:1
isMobile:2
mobileDeviceBranding:1
mobileDeviceModel:1
mobileInputSelector:1
mobileDeviceInfo:1
mobileDeviceMarketingName:1
flashVersion:1
language:1
screenColors:1
screenResolution:1
deviceCategory:3
continent:6
subContinent:22
country:145
region:212
metro:53
city:296
cityId:1
networkDomain:1747
latitude:1
longitude:1
networkLocation:1
campaign:7
source:76
medium:7
keyword:129
isTrueDirect:2
referralPath:198
adContent:16
campaignCode:2
visits:1
hits:96
pageviews:75
bounces:2
newVisits:2
transactionRevenue:147


In [34]:
len(coluna_na)

18

In [35]:
df.drop(coluna_na, axis=1, inplace=True)

In [36]:
df.head()

channelGrouping      date        fullVisitorId  \
0  Organic Search  20160902  9674781571160116268   
1  Organic Search  20160902  8590648239310839049   
2      Affiliates  20160902  9772828344252850532   
3  Organic Search  20160902  1350700416054916432   
4  Organic Search  20160902  1350700416054916432   

                        sessionId     visitId  visitNumber  visitStartTime  \
0  9674781571160116268_1472804607  1472804607            1      1472804607   
1  8590648239310839049_1472835928  1472835928            1      1472835928   
2  9772828344252850532_1472856802  1472856802            1      1472856802   
3  1350700416054916432_1472879649  1472879649            2      1472879649   
4  1350700416054916432_1472829671  1472829671            1      1472829671   

  browser operatingSystem  isMobile  ...         keyword isTrueDirect  \
0  Chrome         Windows     False  ...  (not provided)          NaN   
1  Chrome       Macintosh     False  ...  (not provided)          NaN   
2  Chrome       Macintosh     False  ...             NaN          NaN   
3  Safari             iOS      True  ...  (not provided)         True   
4  Safari             iOS      True  ...  (not provided)          NaN   

  referralPath adContent campaignCode hits pageviews bounces newVisits  \
0          NaN       NaN          NaN    1         1       1         1   
1          NaN       NaN          NaN    1         1       1         1   
2          NaN       NaN          NaN    1         1       1         1   
3          NaN       NaN          NaN    1         1       1       NaN   
4          NaN       NaN          NaN    5         4     NaN         1   

  transactionRevenue  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  

[5 rows x 31 columns]

Usando feature engineering, escolhendo variaveis quantitativas para entender melhor o usuario.

1 - newVisits -> volume de primeira visitas

2 - pageviews -> quantos paginas foi visualizadas

3 - bounces -> rejeicao, visita que gerou apenas um request para o servidor

4 - hits -> volume de interacao que o google recebeu do site 

In [37]:
df.dtypes

channelGrouping       object
date                  object
fullVisitorId         object
sessionId             object
visitId                int64
visitNumber            int64
visitStartTime         int64
browser               object
operatingSystem       object
isMobile                bool
deviceCategory        object
continent             object
subContinent          object
country               object
region                object
metro                 object
city                  object
networkDomain         object
campaign              object
source                object
medium                object
keyword               object
isTrueDirect          object
referralPath          object
adContent             object
campaignCode          object
hits                  object
pageviews             object
bounces               object
newVisits             object
transactionRevenue    object
dtype: object

As variaveis quantitativas precisam ser numeros, as variaveis quantitativas que escolhemos sao do tipo objeto logo precisamos transformar

In [40]:
quant = ['newVisits','pageviews','bounces','hits','transactionRevenue']

In [41]:
for coluna in quant:
    df[coluna] = pd.to_numeric(df[coluna])

In [42]:
df.head()

channelGrouping      date        fullVisitorId  \
0  Organic Search  20160902  9674781571160116268   
1  Organic Search  20160902  8590648239310839049   
2      Affiliates  20160902  9772828344252850532   
3  Organic Search  20160902  1350700416054916432   
4  Organic Search  20160902  1350700416054916432   

                        sessionId     visitId  visitNumber  visitStartTime  \
0  9674781571160116268_1472804607  1472804607            1      1472804607   
1  8590648239310839049_1472835928  1472835928            1      1472835928   
2  9772828344252850532_1472856802  1472856802            1      1472856802   
3  1350700416054916432_1472879649  1472879649            2      1472879649   
4  1350700416054916432_1472829671  1472829671            1      1472829671   

  browser operatingSystem  isMobile  ...         keyword isTrueDirect  \
0  Chrome         Windows     False  ...  (not provided)          NaN   
1  Chrome       Macintosh     False  ...  (not provided)          NaN   
2  Chrome       Macintosh     False  ...             NaN          NaN   
3  Safari             iOS      True  ...  (not provided)         True   
4  Safari             iOS      True  ...  (not provided)          NaN   

  referralPath adContent campaignCode hits pageviews bounces newVisits  \
0          NaN       NaN          NaN    1         1     1.0       1.0   
1          NaN       NaN          NaN    1         1     1.0       1.0   
2          NaN       NaN          NaN    1         1     1.0       1.0   
3          NaN       NaN          NaN    1         1     1.0       NaN   
4          NaN       NaN          NaN    5         4     NaN       1.0   

  transactionRevenue  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  

[5 rows x 31 columns]

Tratando a variavel resposta, ela traz muitos valores vazios


In [43]:
df['transactionRevenue']

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
         ..
12278   NaN
12279   NaN
12280   NaN
12281   NaN
12282   NaN
Name: transactionRevenue, Length: 12283, dtype: float64

O que significa esse valor ser Nan pode ter vários motivos, mas neste caso ele não está preenchido porque o usuário não comprou nada no site.

In [44]:
df['transactionRevenue'].value_counts()

transactionRevenue
13590000.0     3
19190000.0     3
33590000.0     3
31490000.0     2
69990000.0     2
              ..
22490000.0     1
74850000.0     1
32490000.0     1
239880000.0    1
25240000.0     1
Name: count, Length: 146, dtype: int64

In [46]:
df['transactionRevenue'].isna().value_counts()

transactionRevenue
True     12119
False      164
Name: count, dtype: int64

In [48]:
df['transactionRevenue'].isna().sum()

12119

In [49]:
len(df['transactionRevenue'])

12283

Acima podemos ver que o data frame tem um tamanho de 12283 sendo que 12119 são valores NaN, afinal o que deve fazer com esses valores faltantes? Neste caso é interessante preencher com zeros a caso que você deve preencher com média, mediana, moda.


In [51]:
df.transactionRevenue.fillna(0, inplace = True)

In [54]:
print('O valor de zeros: ' + str(df['transactionRevenue'].isna().sum()))


O valor de zeros: 0


Os valores em TrasactionRevenue ou seja o valor que a pessoa gastou no site está em uma formação errada, o valor esta sendo multiplicado por 10^6 ou seja 1000000 então temos que dividir o valor por um milhão.

In [55]:
df.transactionRevenue = df.transactionRevenue / 1000000

In [56]:
df.transactionRevenue.value_counts()

transactionRevenue
0.00     12119
13.59        3
33.59        3
19.19        3
27.19        2
         ...  
24.73        1
6.99         1
14.48        1
95.18        1
25.24        1
Name: count, Length: 147, dtype: int64

Criando variaveis

In [57]:
len(set(df['fullVisitorId']))

9997

In [61]:
df['fullVisitorId'].nunique()

9997

In [64]:
df_quant = df.groupby('fullVisitorId', as_index = False)[quant].sum()

In [65]:
df_quant.head()

fullVisitorId  newVisits  pageviews  bounces  hits  \
0  0002365800130207040        1.0          1      1.0     1   
1  0010286039787739137        1.0          3      0.0     4   
2  0011056874471185769        1.0          2      0.0     2   
3  0014443856125569702        1.0          1      1.0     1   
4  0017260116665815114        0.0          2      0.0     2   

   transactionRevenue  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0

In [66]:
df_quant.shape

(9997, 6)

Treinando o modelo